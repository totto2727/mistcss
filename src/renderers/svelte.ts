import { Data } from '../parser.js'
import { attributeToCamelCase, propertyToCamelCase } from './_case.js'
import { renderPropsInterface, renderTag } from './_common.js'

function renderProps(data: Data): string {
  return [
    'const {',
    [
      ...Object.keys(data.attributes).map(attributeToCamelCase),
      ...Array.from(data.booleanAttributes).map(attributeToCamelCase),
      ...Array.from(data.properties).map(propertyToCamelCase),
      '...props',
    ].join(', '),
    '} = $$props',
  ].join(' ')
}

export function render(filename: string, data: Data): string {
  return `<script lang="ts">
// Generated by MistCSS, do not modify
import type { SvelteHTMLElements } from 'svelte/elements'
import './${filename}.mist.css'

${renderPropsInterface(data, `SvelteHTMLElements['${data.tag}']`)}

type $$Props = Props

${renderProps(data)}
</script>

${renderTag(data, '<slot />', 'class', 'svelte')}
`
}
