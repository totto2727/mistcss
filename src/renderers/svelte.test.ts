import { expect, it, describe } from 'vitest'

import { Data } from '../parser.js'
import { render } from './svelte.js'

describe('render', () => {
  it('renders Svelte component (full)', () => {
    const data: Data = {
      tag: 'div',
      className: 'foo',
      attributes: {
        'data-attr': new Set(['a', 'b']),
        'data-attr-foo-bar': new Set(['foo-bar']),
      },
      booleanAttributes: new Set(['data-is-foo']),
      properties: new Set(['--prop-foo', '--prop-bar']),
    }

    const result = render('component', data)
    expect(result).toMatchSnapshot()
  })

  it('renders Svelte component (minimal)', () => {
    const data: Data = {
      tag: 'div',
      className: 'foo',
      attributes: {},
      booleanAttributes: new Set(),
      properties: new Set(),
    }

    const result = render('component', data)
    expect(result).toMatchSnapshot()
  })

  it('renders Svelte component (void element)', () => {
    const data: Data = {
      tag: 'hr', // hr is a void element and should not have children
      className: 'foo',
      attributes: {},
      booleanAttributes: new Set(),
      properties: new Set(),
    }

    const result = render('component', data)
    expect(result).toMatchSnapshot()
  })
})
