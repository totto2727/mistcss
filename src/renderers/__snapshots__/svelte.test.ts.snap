// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render > renders Svelte component (full) 1`] = `
"<script lang="ts">
// Generated by MistCSS, do not modify
import type { SvelteHTMLElements } from 'svelte/elements'
import './component.mist.css'

type Props = { attr?: 'a' | 'b', attrFooBar?: 'foo-bar', isFoo?: boolean, propFoo?: string, propBar?: string } & SvelteHTMLElements['div']

type $$Props = Props

const { attr, attrFooBar, isFoo, propFoo, propBar, ...props } = $$props
</script>

<div {...props} data-attr={attr} data-attr-foo-bar={attrFooBar} data-is-foo={isFoo} style:--prop-foo={propFoo} style:--prop-bar={propBar} class="foo" ><slot /></div>
"
`;

exports[`render > renders Svelte component (minimal) 1`] = `
"<script lang="ts">
// Generated by MistCSS, do not modify
import type { SvelteHTMLElements } from 'svelte/elements'
import './component.mist.css'

type Props = {  } & SvelteHTMLElements['div']

type $$Props = Props

const { ...props } = $$props
</script>

<div {...props} class="foo" ><slot /></div>
"
`;

exports[`render > renders Svelte component (void element) 1`] = `
"<script lang="ts">
// Generated by MistCSS, do not modify
import type { SvelteHTMLElements } from 'svelte/elements'
import './component.mist.css'

type Props = {  } & SvelteHTMLElements['hr']

type $$Props = Props

const { ...props } = $$props
</script>

<hr {...props} class="foo" />
"
`;
